keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,364 @@
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place
+- [ ] (R) Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [etcd-io/website], for publication to [etcd.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[etcd.io]: https://etcd.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[etcd-io/website]: https://github.com/etcd-io/website
+
+## Summary
+
+We are introducing the Kubernetes style feature gate framework into etcd to gate future feature enhancement behind a sequence of feature lifecyles. Users would be able to turn features on or off and query feature enablement of an etcd server in a consistent way. 
+
+## Motivation
+
+Currently any new enhancements to the etcd are typically added as [an experimental feature](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/features.md#adding-a-new-feature), with a configuration flag prefixed with "experimental", e.g. --experimental-feature-name. 
+
+When it is time to [graduate an experimental feature to stable](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/features.md#graduating-an-experimental-feature-to-stable), a new stable feature flag identical to the experimental feature flag but without the --experimental prefix is added to replace the old feature flag, which is a breaking change.
+
+We are proposing to add [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) in the similar way as Kubernetes does, so that can turn on or off any feature using the `--feature-gates` flag.
+
+Benefits from this enhancement includes:
+
+* graduating a feature is no longer a breaking change with new flags.
+
+* well-established feature lifecycle convention from Kubernetes.
+
+* a unified place to store all the feature enablement state in the code base for simplicity, and easy to refer to in guarding the implementation code.
+
+* easy to query if a feature is available for an ectd server.
```

Comment:
> graduating a feature is no longer a breaking change with new flags.

No. You still need to toggle the gate, which is equivalent to changing command line args.

> well-established feature lifecycle convention from Kubernetes

It works for Kubernetes doesn't necessarily mean it will work for etcd.
Not everything in the world are there to be copied.

> a unified place to store all the feature enablement state in the code base for simplicity, and easy to refer to in guarding the implementation code.

This has nothing to do with feature gates. Users typically don't care how code are organized and we should not force them to.
Exposing flags or configuration items are acceptable when there are user visible changes are introduced.

> easy to query if a feature is available for an ectd server.

When people over engineering something, there always are "reasons".
Please think twice ... do we really need this?

Reply:
> Please think twice ... do we really need this?

Yes, thank you for your feedback.